lib_comfyui/polling_client.py

```diff
@@ -0,0 +1,164 @@
+import asyncio
+import atexit
+import multiprocessing
+import queue
+
+from lib_comfyui.parallel_utils import clear_queue, StoppableThread
+from lib_comfyui import ipc, global_state
+from lib_comfyui.queue_tracker import PromptQueueTracker
+
+
+# rest is ran on comfyui's server side
+class ComfyuiNodeWidgetRequests:
+    start_comfyui_queue = multiprocessing.Queue()
+    finished_comfyui_queue = multiprocessing.Queue()
+    cids = {}
+    param_events = {}
+    last_params = None
+    loop = None
+    focused_key = None
+
+    @ipc.confine_to('comfyui')
+    @staticmethod
+    def send(request_params):
+        clear_queue(ComfyuiNodeWidgetRequests.finished_comfyui_queue)
+        ComfyuiNodeWidgetRequests.start_comfyui_queue.put(request_params)
+        return ComfyuiNodeWidgetRequests.finished_comfyui_queue.get()
+
+    @ipc.confine_to('comfyui')
+    @staticmethod
+    def start_workflow_sync(batch, workflow_type, is_img2img, required_node_types, queue_front):
+        xxx2img = ("img2img" if is_img2img else "txt2img")
+        setattr(global_state, f'{xxx2img}_{workflow_type}_input_images', batch)
+        setattr(global_state, f'tab_name', xxx2img)
+        PromptQueueTracker.update_tracked_id()
+        response = ComfyuiNodeWidgetRequests.send({
+            'request': '/sd-webui-comfyui/webui_request_queue_prompt',
+            'workflowType': f'comfyui_{workflow_type}_{xxx2img}',
+            'requiredNodeTypes': required_node_types,
+            'queueFront': queue_front,
+        })
+
+        if 'error' in response:
+            return response
+
+        # unsafe queue tracking
+        PromptQueueTracker.wait_for_last_put()
+
+        output_key = f'{xxx2img}_{workflow_type}_output_images'
+        results = getattr(global_state, output_key, None)
+
+        if output_key in global_state:
+            delattr(global_state, output_key)
+
+        return results
+
+    @classmethod
+    def set_loop(cls, loop):

```

Comment:
This also starts a listener, which is much more than just setting the loop. There may be a better name for this function

Reply:
idk, `init_request_listener`? Is it necessary to split this function with `_start_listener`

Reply:
it was just for readability. I was intending to only use it as a setter, but I realized that the only moment when I need to start the listener is when the loop has been set. We can rename it to `init_request_listener`